### ISSUE TITLE
Pilot 1.1: concurrent map read and map write - release-1.1-20181109-09-15

### ISSUE THREAD

**Issue by mandarjog** [istio/istio#9916]
Pilot encounters the following error under load. 

`Service.GetServiceAddressForProxy()` reads from `Service.ClusterVIPs` map
and `aggregate/controller.Services()` updates it with new `ClusterID` address.
```
2018-11-12T11:45:43.440637Z     info    ads     LDS: PUSH for node:svc07-0-6-688c999899-nmnpf.service-graph07 addr:"10.40.43.70:58998" listeners:116 207398
fatal error: concurrent map read and map write

goroutine 42888 [running]:
runtime.throw(0x1ad307e, 0x21)
        /usr/local/Cellar/go/1.10/libexec/src/runtime/panic.go:619 +0x81 fp=0xc42e545e20 sp=0xc42e545e00 pc=0x42b381
runtime.mapaccess1_faststr(0x1847d40, 0xc421c41980, 0x1ab0830, 0xa, 0xc421927b88)
        /usr/local/Cellar/go/1.10/libexec/src/runtime/hashmap_fast.go:181 +0x421 fp=0xc42e545e90 sp=0xc42e545e20 pc=0x40ad91
istio.io/istio/pilot/pkg/model.model.Service.GetServiceAddressForProxy(...)
        /Volumes/Untitled/mjog/GOHOME/src/istio.io/istio/pilot/pkg/model/service.go:903
istio.io/istio/pilot/pkg/networking/core/v1alpha3.generateVirtualHostDomains(0xc433194d10, 0x3a9c, 0xc43dd48360, 0x2, 0x2, 0xc431e4f240)
        /Volumes/Untitled/mjog/GOHOME/src/istio.io/istio/pilot/pkg/networking/core/v1alpha3/httproute.go:191 +0x5d2 fp=0xc42e546040 sp=0xc42e545e90 pc=0x14bb8d2
istio.io/istio/pilot/pkg/networking/core/v1alpha3.(*ConfigGeneratorImpl).buildSidecarOutboundHTTPRouteConfig(0xc4201f14a0, 0xc4200dd0e0, 0xc43dd48360, 0xc43b11cb60, 0xc436db8240, 0x1, 0x1, 0xc42908d000, 0x123, 0x200, ...)
        /Volumes/Untitled/mjog/GOHOME/src/istio.io/istio/pilot/pkg/networking/core/v1alpha3/httproute.go:148 +0xbaa fp=0xc42e546740 sp=0xc42e546040 pc=0x14bac7a
istio.io/istio/pilot/pkg/networking/core/v1alpha3.(*ConfigGeneratorImpl).BuildHTTPRoutes(0xc4201f14a0, 0xc4200dd0e0, 0xc43dd48360, 0xc43b11cb60, 0xc425d58134, 0x5, 0xc42000ade0, 0x19d7b60, 0xc433488d60)
```

**Comment by mandarjog** [comment]
@rshriram @costinm

**Comment by hzxuzhonghu** [comment]
/assign 

I can fix it.

**Comment by mandarjog** [comment]
How do you plan to fix it ?

**Comment by mandarjog** [comment]
Adding more locks or using concurrent hash map seem like obvious choices..., but not the best ones.

**Comment by hzxuzhonghu** [comment]
I plan to add a RW lock.

**Comment by mandarjog** [comment]
This is still an issue, I missed a spot in the last code review....
When object is constructed in `buildSidecarOutboundHTTPRouteConfig` the map reference is copied, so there is no unique lock protecting access to the clusterVIP map.

**Comment by mandarjog** [comment]
This is fixed by #10031 and #9952
